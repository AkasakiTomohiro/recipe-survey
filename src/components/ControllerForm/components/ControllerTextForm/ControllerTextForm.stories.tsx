/* eslint-disable react-hooks/rules-of-hooks */
/**
 * @file ControllerTextForm.stories.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { FormProvider, useForm } from 'react-hook-form';
import { z } from 'zod';

import { zodResolver } from '@hookform/resolvers/zod';
import { expect } from '@storybook/jest';
import { userEvent, within } from '@storybook/testing-library';

import { ControllerTextForm } from './ControllerTextForm';

import type { Meta, StoryObj } from '@storybook/react';

type Test = {
  name: string
}

const meta: Meta<typeof ControllerTextForm> = {
  title : 'RecipeSurvey/components/ControllerForm/ControllerTextForm',
  render: (args: any) => {
    const methods = useForm<Test>();
    return (
      <FormProvider {...methods} >
        <ControllerTextForm {...args} />
      </FormProvider>
    );
  }
};

export default meta;

type Story = StoryObj<typeof ControllerTextForm>;

export const Primary: Story = {
  args: {
    label: 'テキスト',
    name : 'name'
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const input = canvas.getByRole('textbox');
    await userEvent.type(input, 'test');
    await expect(input).toHaveValue('test');
  }
};

export const ReadOnly: Story = {
  args: {
    label   : 'テキスト',
    name    : 'name',
    isReadOnly: true
  }
};

export const NumberForm: Story = {
  render: (args: any) => {
    const methods = useForm<{num: number}>();
    return (
      <FormProvider {...methods} >
        <ControllerTextForm {...args} />
      </FormProvider>
    );
  },
  args: {
    label   : 'テキスト',
    name    : 'num',
    type    : 'number',
    onChange: () => {/** */}
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const input = canvas.getByRole('spinbutton');
    await userEvent.type(input, '123');
    await expect(input).toHaveValue(123);
  }
};

const validator = z.object({
  text: z.string().max(1)
}).strict();
export const ErrorTextForm: Story = {
  render: (args: any) => {
    const methods = useForm<z.infer<typeof validator>>({
      mode: 'onChange',
      defaultValues: {
        text: ''
      },
      resolver: zodResolver(validator)
    });
    return (
      <FormProvider {...methods} >
        <ControllerTextForm {...args} />
      </FormProvider>
    );
  },
  args: {
    label               : 'テキスト',
    name                : 'text',
    inputErrorHelperText: "2文字以内で入力してください",
    onChange            : () => {/** */},
    inputProps          : {}
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const input = canvas.getByRole('textbox');
    await userEvent.type(input, 'te');
    await expect(input).toHaveValue("te");
    await expect(canvas.getByText("2文字以内で入力してください")).toBeInTheDocument();
  }
};


