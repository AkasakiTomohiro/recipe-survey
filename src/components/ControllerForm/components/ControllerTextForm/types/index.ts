/**
 * @file index.ts
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { FieldPath } from 'react-hook-form';

import { IGridFormControlProps } from '@/components/ControllerForm';
import { RequiredFilterType } from '@/types';
import { TextFieldProps } from '@mui/material/TextField';

export type IControllerTextFormProps<T extends Record<string, any> = Record<string, any>> = IGridFormControlProps & Omit<TextFieldProps, "variant" | "error"> & {

  /**
   * フォームの名前
   */
   label: string;

   /**
    * controlで管理しているオブジェクトのキーを指定
    */
   name: FieldPath<RequiredFilterType<T, number | string>>;

   /**
    * デフォルトで表示したいヘルパーテキスト
    *
    * @details バリデーションエラーがある場合は表示されない
    */
   inputErrorHelperText?: string;

   /**
    * 編集可能かどうか
    */
   isReadOnly?: boolean;
}