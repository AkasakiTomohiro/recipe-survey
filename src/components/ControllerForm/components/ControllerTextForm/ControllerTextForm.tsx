/**
 * @file ControllerTextForm.tsx
 * @author Artan's Projects
 * @copyright © 2024 Artan's Projects. All rights reserved.
 */

import { useCallback } from 'react';
import { useController, useFormContext } from 'react-hook-form';

import { getZodErrorMessage } from '@/utils/ZodErrorParse';
import TextField from '@mui/material/TextField';

import { GridFormControl } from '../GridFormControl';
import { IControllerTextFormProps } from './types';

/**
 * Primary UI component for user interaction
 *
 * @example
 *
 * 利用したいコンポーネントでジェネリックに型を定義する形でラップして利用すること
 *
 * const TextForm = (props: IControllerTextFormProps<XXXType>) => {
 *   return (<ControllerTextForm {...props}/>);
 * };
 */
export const ControllerTextForm = ({
  name,
  isAllGird,
  onChange,
  inputErrorHelperText,
  isReadOnly,
  ...otherProps
}: IControllerTextFormProps): JSX.Element => {
  const { control, formState: { errors }} = useFormContext();
  const { field } = useController({ name: name, control });

  const errorMessage = getZodErrorMessage(name, errors);

  const onChangeWarp = useCallback((data: React.ChangeEvent<HTMLInputElement>): void => {
    let result: any = data.target.value;
    if(otherProps.type === 'number') {
      result = Number(data.target.value);
    }
    field.onChange({ target: { value: result }});
    if(onChange) {
      onChange(result);
    }
  }, [otherProps.type, field, onChange]);

  return (
    <GridFormControl isAllGird={isAllGird}>
      <TextField
        {...otherProps}
        {...field}
        onChange={onChangeWarp}
        error={Boolean(errorMessage)}
        helperText={Boolean(errorMessage) ? inputErrorHelperText : otherProps.helperText}
        variant="standard"
        inputProps={{ ...(otherProps.inputProps ?? {}), readOnly: isReadOnly }}
      />
    </GridFormControl>
  );
};
