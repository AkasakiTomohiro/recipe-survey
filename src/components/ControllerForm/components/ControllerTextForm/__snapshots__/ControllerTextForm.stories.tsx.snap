// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeSurvey/components/ControllerForm/ControllerTextForm ErrorTextForm play-test 1`] = `
<div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
  <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
    <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
      <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary Mui-error MuiFormLabel-filled Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-zx681a-MuiFormLabel-root-MuiInputLabel-root"
             data-shrink="true"
             for=":r3:"
             id=":r3:-label"
      >
        テキスト
      </label>
      <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-error Mui-focused MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
        <input aria-invalid="true"
               id=":r3:"
               class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
               type="text"
               value="te"
               name="text"
               aria-describedby=":r3:-helper-text"
        >
      </div>
      <p class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium Mui-focused MuiFormHelperText-filled css-10trzr7-MuiFormHelperText-root"
         id=":r3:-helper-text"
      >
        2文字以内で入力してください
      </p>
    </div>
  </div>
</div>
`;

exports[`RecipeSurvey/components/ControllerForm/ControllerTextForm NumberForm play-test 1`] = `
<div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
  <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
    <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
      <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-zx681a-MuiFormLabel-root-MuiInputLabel-root"
             data-shrink="true"
             for=":r2:"
             id=":r2:-label"
      >
        テキスト
      </label>
      <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
        <input aria-invalid="false"
               id=":r2:"
               class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
               type="number"
               name="num"
        >
      </div>
    </div>
  </div>
</div>
`;

exports[`RecipeSurvey/components/ControllerForm/ControllerTextForm Primary play-test 1`] = `
<div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
  <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
    <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
      <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-focused MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-standard css-zx681a-MuiFormLabel-root-MuiInputLabel-root"
             data-shrink="true"
             for=":r0:"
             id=":r0:-label"
      >
        テキスト
      </label>
      <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary Mui-focused MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
        <input aria-invalid="false"
               id=":r0:"
               class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
               type="text"
               name="name"
               value="test"
        >
      </div>
    </div>
  </div>
</div>
`;

exports[`RecipeSurvey/components/ControllerForm/ControllerTextForm ReadOnly smoke-test 1`] = `
<div class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 css-13i4rnv-MuiGrid-root">
  <div class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root">
    <div class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root">
      <label class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-standard css-11mm5xv-MuiFormLabel-root-MuiInputLabel-root"
             data-shrink="false"
             for=":r1:"
             id=":r1:-label"
      >
        テキスト
      </label>
      <div class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-colorPrimary MuiInputBase-formControl css-60qyb9-MuiInputBase-root-MuiInput-root">
        <input aria-invalid="false"
               id=":r1:"
               readonly
               class="MuiInputBase-input MuiInput-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
               type="text"
               name="name"
        >
      </div>
    </div>
  </div>
</div>
`;
